src/V3EmitMk.cpp

```diff
@@ -24,18 +24,514 @@
+
+                if (bucket.m_concatenatedFilenames.size() == 1) {
+                    // Unwrap the bucket if it contains only one file.
+                    FileOrConcatenatedFilesList file;
+                    file.m_fileName = bucket.m_concatenatedFilenames.front();
+                    m_outputFiles.push_back(std::move(file));
+                } else if (bucket.m_concatenatedFilenames.size() > 1) {
+                    m_outputFiles.push_back(std::move(bucket));
+                }
+                // Most likely no bucket will be empty in normal situations. If it happen the
+                // bucket will just be dropped.
+            }
+            for (; fileIt != list.m_files.end(); ++fileIt) {
+                // The Work List is out of buckets, but some files were left.
+                // Add them to the last bucket.
+                m_outputFiles.back().m_concatenatedFilenames.push_back(fileIt->m_filename);
+            }
+        }
+    }
+};
+
 // ######################################################################
 //  Emit statements and expressions
 
 class EmitMk final {
+    using FileOrConcatenatedFilesList = OutputGroup::FileOrConcatenatedFilesList;
+    using FilenameWithScore = OutputGroup::FilenameWithScore;
+
 public:
     // METHODS
 
+    static void emitConcatenatingFile(const FileOrConcatenatedFilesList& entry) {
+        UASSERT(entry.isConcatenatingFile(), "Passed entry does not represent concatenating file");
+
+        V3OutCFile concatenatingFile{v3Global.opt.makeDir() + "/" + entry.m_fileName + ".cpp"};
+        concatenatingFile.putsHeader();
+        for (const auto& file : entry.m_concatenatedFilenames) {
+            concatenatingFile.puts("#include \"" + file + ".cpp\"\n");
+        }
+    }
+
     void putMakeClassEntry(V3OutMkFile& of, const string& name) {
         of.puts("\t" + V3Os::filenameNonDirExt(name) + " \\\n");
     }
 
     void emitClassMake() {
+        std::vector<FileOrConcatenatedFilesList> vmClassesSlowList{};
+        std::vector<FileOrConcatenatedFilesList> vmClassesFastList{};
+        if (v3Global.opt.outputGroups() > 0) {
+            std::vector<FilenameWithScore> slowFiles{};
+            std::vector<FilenameWithScore> fastFiles{};
+            uint64_t slowTotalScore = 0;
+            uint64_t fastTotalScore = 0;
+
+            for (AstNodeFile* nodep = v3Global.rootp()->filesp(); nodep;
+                 nodep = VN_AS(nodep->nextp(), NodeFile)) {
+                const AstCFile* const cfilep = VN_CAST(nodep, CFile);
+                if (cfilep && cfilep->source() && cfilep->support() == false) {
+                    std::vector<FilenameWithScore>& files = cfilep->slow() ? slowFiles : fastFiles;
+                    uint64_t& totalScore = cfilep->slow() ? slowTotalScore : fastTotalScore;
+
+                    FilenameWithScore f;
+                    f.m_filename = V3Os::filenameNonDirExt(cfilep->name());
+                    f.m_score = cfilep->complexityScore();
```

Comment:
```suggestion
                    FilenameWithScore f{V3Os::filenameNonDirExt(cfilep->name(), cfilep->complexityScore()};
```

And m_filename in the struct can probably become const.